File: java/src/test/java/ev/projects/GameEndingTest.java
package ev.projects;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class GameEndingTest {

    @Test
    void playerScoresForCompletedHorizontalLines() {
        Player player = new PlayerMother().newPlayer(new WallMother().withCompletedTwoHorizontalLines());
        Game game = new GameMother().new2PlayerGame(player);
        int currentScore = player.score();

        game.executeGameEndingPhase();

        assertEquals(4, player.score() - currentScore);
    }

    @Test
    void playerScoresForCompletedVerticalLines() {
        Player player = new PlayerMother().newPlayer(new WallMother().withThreeCompletedVerticalLines());
        Game game = new GameMother().new2PlayerGame(player);
        int currentScore = player.score();

        game.executeGameEndingPhase();

        assertEquals(21, player.score() - currentScore);
    }

    @Test
    void playerScoresForCompletedTiles() {
        Player player = new PlayerMother().newPlayer(new WallMother().withCompletedBlueAndYellowTiles());
        Game game = new GameMother().new2PlayerGame(player);
        int currentScore = player.score();

        game.executeGameEndingPhase();

        assertEquals(20, player.score() - currentScore);
    }

    @Test
    void winnerIsDeclared() {
        Player player1 = new PlayerMother().newPlayer("Joke");
        player1.addScore(10);
        Player player2 = new PlayerMother().newPlayer("Alfonso");
        player2.addScore(20);
        Player player3 = new PlayerMother().newPlayer("Ra");
        player3.addScore(15);
        Game game = new Game(List.of(player1, player2, player3));

        List<String> winner = game.winners();

        assertEquals(1, winner.size());
        assertEquals("Alfonso", winner.get(0));
    }

    @Test
    void resolvesDrawWithHorizontalLines() {
        Player player1 = new PlayerMother().newPlayer(new WallMother().withCompletedTwoHorizontalLines(), "Joke");
        player1.addScore(20);
        Player player2 = new PlayerMother().newPlayer(new WallMother().withCompletedHorizontalLine(), "Alfonso");
        player2.addScore(20);
        Game game = new Game(List.of(player1, player2));

        List<String> winner = game.winners();

        assertEquals(1, winner.size());
        assertEquals("Joke", winner.get(0));
    }

    @Test
    void bothPlayersWinIfDrawnAndSameAmountOfCompletedHorizontalLines() {
        Player player1 = new PlayerMother().newPlayer(new WallMother().withCompletedTwoHorizontalLines(), "Joke");
        player1.addScore(20);
        Player player2 = new PlayerMother().newPlayer(new WallMother().withCompletedTwoHorizontalLines(), "Alfonso");
        player2.addScore(20);
        Game game = new Game(List.of(player1, player2));

        List<String> winners = game.winners();

        assertEquals(2, winners.size());
    }
}

File: python/test/game_ending_test.py
import unittest
from flask import Flask
from azul_ai_gym.game import Game
from wall_mother import WallMother
from game_mother import GameMother
from player_mother import PlayerMother


class TestGameEnding(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.app = Flask(__name__)
        cls.app_context = cls.app.app_context()
        cls.app_context.push()

    @classmethod
    def tearDownClass(cls):
        cls.app_context.pop()

    def test_player_scores_for_completed_horizontal_lines(self):
        player = PlayerMother().new_player(wall=WallMother().with_completed_two_horizontal_lines())
        game = GameMother.new_2_player_game(player1=player)
        current_score = player.score

        game.execute_game_ending_phase()

        self.assertEqual(4, player.score - current_score)

    def test_player_scores_for_completed_vertical_lines(self):
        player = PlayerMother().new_player(wall=WallMother().with_three_completed_vertical_lines())
        game = GameMother.new_2_player_game(player1=player)
        current_score = player.score

        game.execute_game_ending_phase()

        self.assertEqual(21, player.score - current_score)

    def test_player_scores_for_completed_tiles(self):
        player = PlayerMother().new_player(wall=WallMother().with_completed_blue_and_yellow_tiles())
        game = GameMother.new_2_player_game(player1=player)
        current_score = player.score

        game.execute_game_ending_phase()

        self.assertEqual(20, player.score - current_score)

    def test_winner_is_declared(self):
        player1 = PlayerMother().new_player(name="Joke")
        player1.add_score(10)
        player2 = PlayerMother().new_player(name="Alfonso")
        player2.add_score(20)
        player3 = PlayerMother().new_player(name="Ra")
        player3.add_score(15)
        game = Game([player1, player2, player3])

        winner = game.winners()

        self.assertEqual(1, len(winner))
        self.assertEqual("Alfonso", winner[0])

    def test_resolves_draw_with_horizontal_lines(self):
        player1 = PlayerMother().new_player(wall=WallMother().with_completed_two_horizontal_lines(), name="Joke")
        player1.add_score(20)
        player2 = PlayerMother().new_player(wall=WallMother().with_completed_horizontal_line(), name="Alfonso")
        player2.add_score(20)
        game = Game([player1, player2])

        winner = game.winners()

        self.assertEqual(1, len(winner))
        self.assertEqual("Joke", winner[0])

    def test_both_players_win_if_drawn_and_same_amount_of_completed_horizontal_lines(self):
        player1 = PlayerMother().new_player(wall=WallMother().with_completed_two_horizontal_lines(), name="Joke")
        player1.add_score(20)
        player2 = PlayerMother().new_player(wall=WallMother().with_completed_two_horizontal_lines(), name="Alfonso")
        player2.add_score(20)
        game = Game([player1, player2])

        winners = game.winners()

        self.assertEqual(2, len(winners))